opentelemetry-sdk/src/logs/log_processor.rs

```diff
@@ -292,87 +291,112 @@ impl Debug for BatchLogProcessor {
+                            }
                         }
                     }
                 }
             }
+            Err(mpsc::TrySendError::Full(_)) => {
+                // Increment dropped logs count. The first time we have to drop
+                // a log, emit a warning.
+                if self.dropped_logs_count.fetch_add(1, Ordering::Relaxed) == 0 {
+                    otel_warn!(name: "BatchLogProcessor.LogDroppingStarted",
+                        message = "BatchLogProcessor dropped a LogRecord due to queue full/internal errors. No further log will be emitted for further drops until Shutdown. During Shutdown time, a log will be emitted with exact count of total logs dropped.");
+                }
+            }
+            Err(mpsc::TrySendError::Disconnected(_)) => {
+                // Given background thread is the only receiver, and it's
+                // disconnected, it indicates the thread is shutdown
+                otel_warn!(
+                    name: "BatchLogProcessor.Emit.AfterShutdown",
+                    message = "Logs are being emitted even after Shutdown. This indicates incorrect lifecycle management of OTelLoggerProvider in application. Logs will not be exported."
+                );
+            }
         }
     }
 
     fn force_flush(&self) -> LogResult<()> {
-        if self.is_shutdown.load(std::sync::atomic::Ordering::Relaxed) {
-            return LogResult::Err(LogError::Other(
-                "BatchLogProcessor is already shutdown".into(),
-            ));
-        }
         let (sender, receiver) = mpsc::sync_channel(1);
-        self.message_sender
+        match self
+            .message_sender
             .try_send(BatchMessage::ForceFlush(sender))
-            .map_err(|err| LogError::Other(err.into()))?;
-
-        receiver
-            .recv_timeout(self.forceflush_timeout)
-            .map_err(|err| {
-                if err == RecvTimeoutError::Timeout {
-                    LogError::ExportTimedOut(self.forceflush_timeout)
-                } else {
-                    LogError::Other(err.into())
-                }
-            })?
+        {
+            Ok(_) => receiver
+                .recv_timeout(self.forceflush_timeout)
+                .map_err(|err| {
+                    if err == RecvTimeoutError::Timeout {
+                        LogError::ExportTimedOut(self.forceflush_timeout)
+                    } else {
+                        LogError::Other(err.into())
+                    }
+                })?,
+            Err(mpsc::TrySendError::Full(_)) => {
+                // If the control message could not be sent, emit a warning.
+                otel_debug!(
+                    name: "BatchLogProcessor.ForceFlush.ControlChannelFull",
+                    message = "Control message to flush the worker thread could not be sent as the control channel is full. This can occur if user repeatedily calls force_flush without finishing the previous call."
```

Comment:
I think can also happen if user repeatedly calls the shutdown message, and the first shutdown message is not yet processed, so channel is still intact :)

Reply:
that is a good point. Let me modify error message to include that.

Reply:
will do in separate PR to keep this merged asap.